pandas/io/parsers/python_parser.py

```diff
@@ -1189,21 +1189,35 @@ def _rows_to_cols(self, content: list[list[Scalar]]) -> list[np.ndarray]:
 
             for i, _content in iter_content:
                 actual_len = len(_content)
-
                 if actual_len > col_len:
                     if callable(self.on_bad_lines):
                         new_l = self.on_bad_lines(_content)
                         if new_l is not None:
-                            content.append(new_l)  # pyright: ignore[reportArgumentType]
+                            # Truncate extra elements and warn.
+                            new_l = cast(list[Scalar], new_l)
+                            if len(new_l) > col_len:
+                                warnings.warn(
+                                    "Length of header or names does not match length "
+                                    "of data. This leads "
+                                    "to a loss of data with index_col=False.",
+                                    ParserWarning,
+                                    stacklevel=find_stack_level(),
```

Comment:
I'm not a fan of warning here in addition to `alert_malformed`. Can you use whether `self.on_bad_lines` is callable to determine a better warning in this case (e.g. "Length from `on_bad_lines` callable does not match..."), and then raise the appropriate warning in `_alert_malformed`.